proposed/2023/package-source-mapping-actions-PMUI.md

```diff
@@ -0,0 +1,102 @@
+
+## Explanation
+
+### Functional Explanation
+
+When a new package source mapping is being created, the Preview Window will appear showing which packages (the top-level and any transitive dependencies) will be mapped. 
+
+(TBD) It is not yet understood whether there will be a new afforance on the details pane in a new row below the Install/Update buttons, or if source mappings will be handled in the background.
+
+#### A Proposal for Consent to a new Source Mapping
+ A `CheckBox` that allows the customer to choose if they want to add a mapping to the package from the selected source when they `Install`/`Update`. This `CheckBox` will always be shown even if package source mapping is not enabled. Package source mapping can be enabled by the customer checking the `CheckBox` to create the first mapping for the solution's settings. Once package source mapping is enabled, the customer must either have a mapping previously configured, or consent to a new mapping by checking the `CheckBox` to install. Otherwise, Restore will fail as it does today.
+
+ PSM Enabled|Package Mapped|Current Behavior|Proposed Behavior
+---|---|------|------|
+No|No|regular install|regular install
+Yes|No|Error in Error List|Preview Window appears showing mappings being created to selected source
+Yes|Yes|regular install|Allow creating an additional source mapping to selected source for each newly installed package (including transitives)
+ 
+ The existing Preview Window dialog that shows when the customer hits `Install` or `Update` will be modified to list the mappings that will be made by their action. The customer already has the option to disable this preview dialog, so if they choose to disable it, there will be nothing telling them what mappings they made. 
+ 
+ A proposal for the changes to the preview window:
+![PM UI - a proposal for the changes to the preview window](../../meta/resources/PackageSourceMapping/PMUI_preview.png)
+
+#### Restore Errors
+
+There are a few errors the customer could make while trying to install or update a package with a mapping that would make `Restore` fail. First, there could be issues with the transitive packages. Maybe there are transitive packages that are already mapped to a different source or the source that the customer is trying to map a package to does not support some of the transitive packages. 
+
+The customer could also try to update a package that is already mapped to a different source, but the source it is mapped to does not support the version they are trying to update to. In these scenarios, `Restore` will fail in the same way it does today. The customer can see why `Restore` failed in the output window. 
+
+Currently, `NU1100` is used for both an incompatible TFM, and for a missing Package Source Mapping. 
```

Comment:
I think I mentioned in the issue, but NU1100 is always a missing package. 

Missing package for source mapping vs not shouldn't matter. This was a deliberate decision at that point. 

Ofc it can be a subject to change, but I think this particular statement is not correct.

Reply:
Yes, it's missing, but the suggestion was to be more specific that it's missing *because* package source mapping didn't allow restore to retrieve it.

Here, the NU1100 is shown only because I deleted the mapping for the package I tried to install:
![image](https://github.com/NuGet/Home/assets/49205731/d38a4410-fb77-42e4-93a0-abfde7309386)

Reply:
So NU1100 says, `PackageSourceMapping is enabled, the following source(s) were not considered: X, Y`. 

How would you make that clearer?

Reply:
The proposal was for me to create something distinct. Copying from the Desired behavior in https://github.com/NuGet/Home/issues/12606:

> Specific messages for either scenario.
> Specify whether the TFM not compatible, or is the PackageSourceMapping not configured?

Reply:
My point is that we already have different text for source mapping vs no source mapping scenarios. Are you proposing a new error code? This was something we considered and *specifically* opted against that since when a package is not resolved, the issue could be, package doesn't exist, package exists but version doesn't, no suitable version can be selected etc. All of those get the same "due to source mapping, these sources were not considered". You can look at that in https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Commands/RestoreCommand/Diagnostics/UnresolvedMessages.cs by searching for `isPackageSourceMappingEnabled `.

For the 2nd line, compatibility has different warning/error codes, NU1100s vs NU1200s. In fact they are generated at different points, since in the first case there is no package, but in the 2nd case, there's a package that *was* analyzed.

Reply:
Hm, when Svetlana and I were walking through the changes to PM UI, I'm sure we both saw a case where the package *was* available on a source, but we got a NU1100.  I'll have to see if I can recreate that. 

According to the docs, TargetFramework is called out as well:

> Unable to resolve 'Dependency (>= 1.0.0)' for 'TargetFramework'

https://learn.microsoft.com/en-us/nuget/reference/errors-and-warnings/nu1100

Reply:
> According to the docs, TargetFramework is called out as well:

It's calling that out not because it is a compatibility problem, but because projects can multi target.